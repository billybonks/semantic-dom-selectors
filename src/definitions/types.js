export let button = [
  'button',
  'a',
  '[role="button"]',
  'input[type="reset"]',
  'input[type="button"]',
  'input[type="submit"]',
  '[role="link"]',
  '[role="menuitem"]',
  'input[alt][type="image"]'
];


export let text = [
  'input',
  'textarea',
  '[role="slider"]',
  '[role="spinbutton"]',
  '[role="textbox"]',
  '[contenteditable="true"]',
]

export let toggle = [
  '[role="menuitemcheckbox"]',
  '[role="checkbox"]',
  '[type="checkbox"]',
]

export let select = [
  '[role="menuitemradio"]',
  'select',
  '[role="listbox"]',
  '[role="radiogroup"]',
]
