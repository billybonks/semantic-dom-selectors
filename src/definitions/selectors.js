import {
  button, text, toggle, select,
} from './types';

export const buttonQuery = button.join(',');
export const textQuery = text.join(',');
export const toggleQuery = toggle.join(',');
export const selectQuery = select.join(',');
export const formControlQuery = [textQuery, toggleQuery, selectQuery].join(',');
