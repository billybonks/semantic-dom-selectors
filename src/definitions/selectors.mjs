import { button, text, toggle, select } from './types';

export let buttonQuery = button.join(',');
export let textQuery  = text.join(',');
export let toggleQuery = toggle.join(',');
export let selectQuery = select.join(',');
export let formControlQuery = [textQuery, toggleQuery, selectQuery].join(',');
