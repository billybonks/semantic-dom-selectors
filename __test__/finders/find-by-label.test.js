import findByLabel from '../../src/finders/find-by-label';
import { textQuery } from '../../src/definitions/selectors';

describe('Find by label', () => {
  afterEach(() => {
    document.body.innerHTML = '';
  });

  describe('Label exists', () => {
    afterEach(() => {
      document.body.innerHTML = '';
    });

    describe('with "for" attribute targetting input', () => {
      beforeEach(() => {
        const markup = `
          <label for="proper">Hello!</label>
          <label for="notproper">Another Hello!</label>
          <div id="proper">
            <input name="proper" value="happy" />
          </div>
          <input name="notproper" value="sad"/>
        `;
        document.body.innerHTML = markup;
      });

      it('finds the right input', () => {
        const expectedInput = document.querySelector('input[value="happy"]');
        const foundInput = findByLabel.run(textQuery, 'Hello!');
        expect(foundInput.length).toEqual(1);
        expect(foundInput[0]).toEqual(expectedInput);
      });
    });

    describe('with "for" attribute targetting div', () => {
      beforeEach(() => {
        const markup = `
          <label for="control">Label of control</label>
          <div class="day-slider">
            <div id="control" class="day-slider-handle" role="slider"
               aria-valuemin="1"
               aria-valuemax="7"
               aria-valuenow="2"
               aria-valuetext="Monday">
           </div>
          </div>
        `;
        document.body.innerHTML = markup;
      });

      it('finds the right input', () => {
        const expectedInput = document.querySelector('#control');
        const foundInput = findByLabel.run(textQuery, 'Label of control');
        expect(foundInput.length).toEqual(1);
        expect(foundInput[0]).toEqual(expectedInput);
      });
    });

    describe('without "for" attribute', () => {
      beforeEach(() => {
        const markup = `
          <label>Hello!</label>
        `;
        document.body.innerHTML = markup;
      });

      it('throws error', async () => {
        expect(() => {
          findByLabel.run(textQuery, 'Hello!');
        }).toThrowErrorMatchingSnapshot();
      });
    });
  });

  describe('Label does not exist', () => {
    beforeEach(() => {
      const markup = `
        <div for="trickery">Hello!</div>
        <input name="trickery"/>
      `;
      document.body.innerHTML = markup;
    });

    it('returns null', () => {
      expect(findByLabel.run(textQuery, 'Hello!')).toEqual(null);
    });
  });
});
