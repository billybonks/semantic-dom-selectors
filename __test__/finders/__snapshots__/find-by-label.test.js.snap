// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Find by label Label exists without "for" attribute throws error 1`] = `"Label was found for Hello! but it did not have a for attribute"`;
